src/operator/numpy/linalg/np_gesvd-inl.h

```diff
@@ -0,0 +1,298 @@
+                         const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  CHECK_EQ(inputs.size(), 1);
+  CHECK_EQ(outputs.size(), 3);
+  if (inputs[0].shape_.Size() == 0) {
+    return;
+  }
+  MSHADOW_SGL_DBL_TYPE_SWITCH(outputs[0].type_flag_, OType, {
+    mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+    laop::op(inputs[0].FlatToKD<xpu, 3, OType>(s),
+             outputs[0].FlatToKD<xpu, 3, OType>(s),
+             outputs[1].FlatToKD<xpu, 2, OType>(s),
+             outputs[2].FlatToKD<xpu, 3, OType>(s), ctx, attrs);
+  });
+}
+
+// Helper for gesvd_backward. See technical report for details
+template<typename DType>
+DType gesvd_back_helper_eps(DType* X);
+
+template<>
+MSHADOW_XINLINE float gesvd_back_helper_eps(float* X) {
+  return 1e-30;
+}
+
+template<>
+MSHADOW_XINLINE double gesvd_back_helper_eps(double* X) {
+  return 1e-100;
+}
+
+struct GesvdBackHelper_dV {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int k, int m, int n, DType* L, int ldl,
+                                  DType* dA, int ldda) {
+    const int offl(k * ldl);
+    const int offda(k * m * ldda);
+    DType denom(0.0);
+    const DType eps(gesvd_back_helper_eps(dA));
+    for (int i = 0; i < m; ++i) {
+      denom = L[offl + i];
+      if (denom < eps) denom = eps;
+      for (int j = 0; j < n; ++j) {
+        dA[offda + i * ldda + j] /= denom;
+      }
+    }
+  }
+};
+
+struct GesvdBackHelper_G1 {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int k, int m, int n, DType* X, int ldx,
+                                  DType* L, int ldl) {
+    const int offl(k * ldl);
+    const int offx(k * m * ldx);
+    DType numer(0.0);
+    for (int i = 0; i < m; ++i) {
+      for (int j = 0; j < m; ++j) {
+        numer = L[offl + j];
+        X[offx + i * ldx + j] *= numer;
+      }
+    }
+  }
+};
+
+struct GesvdBackHelper_G2 {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int k, int m, int n, DType* X, int ldx,
+                                  DType* L, int ldl, DType* dL, int lddl,
+                                  DType* dA, int ldda, DType* V, int ldv) {
+    const int offx(k * m * ldx);
+    const int offl(k * ldl);
+    const int offdl(k * lddl);
+    const int offda(k * m * ldda);
+    const int offv(k * m * ldv);
+    const DType eps(gesvd_back_helper_eps(X));
+    DType denom1(0.0), denom2(0.0), elem(0.0);
+
+    for (int i = 0; i < m - 1; ++i) {
+      for (int j = i + 1; j < m; ++j) {
+        denom1 = L[offl + i] - L[offl + j];
+        denom2 = L[offl + i] + L[offl + j];
+        if (denom1 < eps) denom1 = eps;
+        if (denom2 < eps) denom2 = eps;
+        elem = (X[offx + i * ldx + j] - X[offx + j * ldx + i]) / denom1 / denom2;
+        X[offx + i * ldx + j] = elem * L[offl + j];
+        X[offx + j * ldx + i] = elem * L[offl + i];
+      }
+    }
+    for (int i = 0; i < m; ++i) {
+      elem = DType(0.0);
+      for (int j = 0; j < n; ++j) {
```

Comment:
This is not needed. You have computed this before already, using gemm, just pass in the diagonal.
Also, the function does not need dA and V as input args.

Reply:
Yes, fixed.